tests/lib/compose_features.py

```diff
@@ -0,0 +1,273 @@
+        dir) if os.path.isfile(os.path.join(dir, f))]
+    originals, reruns = _get_original_and_rerun_list(filenames)
+    for rerun in reruns:
+        result_name = _get_name_without_run_number(rerun)
+        original = list(
+            filter(lambda x: x.startswith(result_name), originals))
+        if len(original) != 1:
+            raise RuntimeError(
+                f'The rerun {rerun} does not have a corresponding original run')
+        tests = _replace_tests(os.path.join(
+            dir, original[0]), os.path.join(dir, rerun))
+        with open(os.path.join(output_dir, result_name + '.json'), 'w') as f:
+            f.write(json.dumps(tests))
+
+    # Search for system test results that had no reruns and
+    # simply copy their result file to the output folder
+    for file in filenames:
+        if file not in originals and file not in reruns:
+            shutil.copyfile(os.path.join(dir, file),
+                            os.path.join(output_dir, _get_name_without_run_number(file) + '.json'))
+
+
+def run_attempt_type(value: Any) -> Any:
+    if value is not int or int(value) <= 0:
+        raise argparse.ArgumentTypeError(
+            f'{value} is invalid. Run attempts are integers and start at 1')
+    return value
+
+
+if __name__ == '__main__':
+    description = '''
+    Can be run in two modes: composed feature generation or composed feature consolidation
+
+    Composed feature generation mode
+
+    Given a directory containing files with outputs of journal-analzyer.py with filenames
+    of format <backend>:<system>:suite--path--<test>:<variant>, it will construct a json
+    file for each <backend>:<system> with feature-tagging information, accompanied with
+    additional test information.
+
+    Composed feature consolidation mode
+
+    Given a directory containing files of pre-composed feature information with filenames like
+    <backend>:<system>_<run-attempt>.json, it writes the consolidated feature information in a
+    new directory (specified with the --output flag) where the latest rerun data replaces the old.
+    So if a file contains one test that was later rerun, the new consolidated file will contain
+    unaltered content from the original run except for the one test rerun that will replace
+    the old.
+    '''
+    parser = argparse.ArgumentParser(
+        description=description, formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('-d', '--dir', type=str, required=True,
+                        help='Path to the folder containing json files')
+    parser.add_argument('-o', '--output', type=str,
+                        help='Output directory', required=True)
+    parser.add_argument('-s', '--scenarios', type=str,
+                        help='Comma-separated list of scenarios', default='')
+    parser.add_argument('-e', '--env-variables', type=str,
+                        help='Comma-separated list of environment variables as key=value', default='')
```

Comment:
Nitpick, what if the value needs to contain a comma?

You can consider to use `narg='*'` and build a list of key-value pairs.

Second nitpick, if you pass `type=` you can write a simple class like, the one below to both handle value checks before rest of main runs, *and* provide typed data earlier, so that fewer things try to parse strings the right way at the right time.
```python
class EnvPair:
  key: str
  value: str

  def __init__(self, s: str) -> None:
    self.key, eq, self.value = s.partition('=')
    if eq != '-':
      raise ValueError("not a key=value pair")
```